### ISSUE TITLE
[HOLD for payment 2022-09-16] [$1000] - Unable to add attachment or request money using keyboard navigation

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7317]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. With cursor in compose box Type cmd+tab to navigate to :heavy_plus_sign:
3. Tap return to open Request Money and Add Attachment menu
4. Attempt to select one via the arrow keys

## Expected Result:
Ability to navigate to  Request Money or Add Attachment and to tap return to select.

## Actual Result:
Can not navigate with arrow keys nor select Request Money or Add Attachment

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Desktop App

**Version Number:** 1.1.30-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/150068401-0b9222c7-cd19-4fcf-b979-60b2239cc05b.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1642453579084400

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alex-mechler (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by alex-mechler** [comment]
Frontend only issue, can easily go external! Thanks for the report @mallenexpensify

**Comment by MelvinBot** [comment]
Triggered auto assignment to @bfitzexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by bfitzexpensify** [comment]
Upwork post [here](https://www.upwork.com/jobs/~0140b54fd42619da64)

**Comment by nlundgren** [comment]
Hi there, 

I am here routed from Upwork to make a proposal for this issue. My implementation would consist on keyboard listeners that trigger specifically when that menu is open. 

The steps to solve this are:
- Identify if menu (request/attachment) is open. 
- When it is open add keyboard listeners in order to change the active status of each menu item.
- Another listener on key enter will activate the desired tab while the menu is opened, and close it after it is pressed.

Best,

Nils

**Comment by parasharrajat** [comment]
@nlundgren Thanks for your proposal. Although the steps are clear, you need to add more information. 
1. Where will you make these changes? 
2. How will you do that? 
3. Code samples?

**Comment by nlundgren** [comment]
These changes will be made on the on the chat view. I will create a handleKeyboardInput() function and pass it down to the chat component. Once the user opens the + sign while using the keyboard I will use this function to perform the mentioned actions. 

In order to know exactly where, I need to spend some more time learning the structure of the project. The main components I saw when going over it are from the React Native application, as I understand this is an accessibility issue in the web app. I would happily spend the necessary time on it if we create a project proposal through Upwork and establish some deliverables.

**Comment by parasharrajat** [comment]
Ok. I would suggest you take your time to understand the codebase and layout a proper proposal to the problem. 

You can head over to README.md file to understand the project structure and contribution guidelines. 

> Copied from one of core team member

Our approach on all our open source issues is to first discuss the solution with the contributors in order to make sure we all agree on the solution and that it is the best solution we can find. So we do not hire anyone before we can evaluate the solution. 
I know this is in contrast to other job offers published in upwork, but it allows us to ensure the level of quality we want for our app and code while also diminishing the work thrown out by contributors. ie: if we waited till contributors submit a PR and then found out the solution is not acceptable or has some drawbacks, then that code would need to be thrown out, which is frustrating for contributors.

**Comment by bfitzexpensify** [comment]
Not much action here so far — doubled the price to $500

**Comment by bfitzexpensify** [comment]
No progress again for 7 days, doubling again to $1000. Upwork [job](https://www.upwork.com/jobs/~0140b54fd42619da64)

**Comment by mananjadhav** [comment]
**Proposal**
I'll add a `keyboard` handler to `BasePopoverMenu`. Can be optionally added by a prop so that Keyboard events are listened only based on the prop say `allowKeyboardMovement`.

1. We store the active menu in the state: `this.state = { activeMenuIndex: -1 }`
2. On `componentDidMount`, we set up the events
```
if (keyBoardEvent.key.startsWith('Arrow')) {
    this.highlightActiveMenu(keyBoardEvent.key);
    return;
}
if (keyBoardEvent.key === 'Enter' && this.state.activeMenuIndex !== -1) {
    this.props.onItemSelected(this.props.menuItems[this.state.activeMenuIndex]);
}
```

3. We add `cleanUpHandlers`, which will remove keyboard handlers on componentWillUnmount.
4. Our `highlightActiveMenu` will look like this:
 ```
highlightActiveMenu(arrowKey) {
        console.log('Activate', arrowKey);
        let updatedActiveMenuIndex = this.state.activeMenuIndex;
        if (arrowKey !== 'ArrowDown' && arrowKey !== 'ArrowUp') {
            return;
        }


        if (arrowKey === 'ArrowDown') {
            if (updatedActiveMenuIndex === -1 || updatedActiveMenuIndex === this.props.menuItems.length - 1) {
                updatedActiveMenuIndex = 0;
            } else {
                updatedActiveMenuIndex += 1;
            }
        }

        if (arrowKey === 'ArrowUp') {
            if (updatedActiveMenuIndex === -1 || updatedActiveMenuIndex === 0) {
                updatedActiveMenuIndex = this.props.menuItems.length - 1;
            } else {
                updatedActiveMenuIndex -= 1;
            }
        }

        console.log('Updated Value', updatedActiveMenuIndex);
        this.setState({activeMenuIndex: updatedActiveMenuIndex});
    }


```
5. We add a `focused={index === this.state.activeMenuIndex}` prop to `MenuItems`

**Comment by parasharrajat** [comment]
@mananjadhav 's Proposal looks good to me. I like that it is generic and can be applied to other menus easily.

A few points: 
1. Menu should be highlighted with a blue border to indicate focus like TextInput.

cc: @puneetlath 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by puneetlath** [comment]
That proposal also seems good to me. 

@stitesExpensify are we ok going forward with this or is this being handled as part of the broader accessibility project?

**Comment by stitesExpensify** [comment]
Hmm @mvtglobally would arrow keys be the way that most visually/mobility impaired users try to access that menu (as opposed to something like tabbing through)? cc: @ogumen

**Comment by ogumen** [comment]
> Hmm @mvtglobally would arrow keys be the way that most visually/mobility impaired users try to access that menu (as opposed to something like tabbing through)? cc: @ogumen

Up/down arrow keys (and the circular navigation) is the expected keyboard navigation pattern for the menu options. Pressing Tab key when the popover menu is expanded should collapse it and move focus to the next interactive element (visually it is the "Write something" text field). A good example of the popover menu implementation can be found here: https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-links.html  It is for web, but you can check the keyboard interaction.

**Comment by mvtglobally** [comment]
@stitesExpensify i would probably have to defer to @ogumen to confirm the best practices when it comes to accessibility issues

**Comment by puneetlath** [comment]
Ok cool. In that case @mananjadhav's proposal sounds good to me too.

**Comment by MelvinBot** [comment]
📣 @mananjadhav You have been assigned to this job by @puneetlath!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by mananjadhav** [comment]
Was a bit unwell, will push an update tomorrow.

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the Exported assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mountiny (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by puneetlath** [comment]
@mountiny the PR is currently on hold, but I've assigned you to both this issue and [the PR](https://github.com/Expensify/App/pull/7846) since I'm going on sabbatical for two months. Thanks!

**Comment by mountiny** [comment]
Removing the `Help Wanted` label left by adding the `Exported` label again.

**Comment by parasharrajat** [comment]
PR is no more on HOLD. Looking forward to updates.

**Comment by mananjadhav** [comment]
Completing it this weekend

**Comment by parasharrajat** [comment]
Curious to know where are we with this.

**Comment by mananjadhav** [comment]
I am a little stuck here. I tried setting the active menu with onFocus, but it doesn't work well. `focused` item still adds a double border. Flicker is removed. 

I also tried listening to the TAB key, but the circular navigation is breaking. I'll update as soon as I have it working.

**Comment by mountiny** [comment]
@mananjadhav No worries and thanks for the update. Glad to have you back 🙏

**Comment by mananjadhav** [comment]
Thanks, @mountiny. Apologies might need 1-2 days to get to close it.

**Comment by mountiny** [comment]
I am going to be ooo for 5 weeks so I am reassigning this issue to a new CME. I will assign you to the PR review too! Thank you very much 🙇

**Comment by bfitzexpensify** [comment]
Hi @mananjadhav — how are things looking here? Anything that you need help with?

**Comment by mananjadhav** [comment]
No help. I answered some questions on the feedback and one question.

**Comment by marcaaron** [comment]
Pulling a @mountiny and assigning someone new as I'm going OOO for a week.

**Comment by bfitzexpensify** [comment]
HI @mananjadhav — any update here?

**Comment by mananjadhav** [comment]
@bfitzexpensify On hold, till we merge #7702. https://github.com/Expensify/App/pull/7846#issuecomment-1119652363

I've updated the title of the PR. Can you update the title of this issue?

**Comment by mountiny** [comment]
Taking back over from Alex, thanks for help!

**Comment by mananjadhav** [comment]
@roryabraham Are we hoping to close #7702 by next week?

**Comment by parasharrajat** [comment]
Great news #7702 is done. There could be bugs but we should be good to restart this issue.

**Comment by mallenexpensify** [comment]
Should we wait for the #7702 to hit production first @parasharrajat / @mananjadhav ?

**Comment by parasharrajat** [comment]
I don't mind waiting.

**Comment by parasharrajat** [comment]
@mananjadhav let's complete this PR.

**Comment by mananjadhav** [comment]
Was finishing the Skeleton UI until yesterday. Will start with this tomorrow.

**Comment by mountiny** [comment]
@mananjadhav How's it looking? Any estimate time for the PR?

**Comment by mananjadhav** [comment]
Will be ready by tomorrow. I am still a bit stuck up in adding signed commits for the Skeleton UI

**Comment by mananjadhav** [comment]
I've finally started with this today. I just managed to fix the Skeleton UI issues and reworked that PR until last week.

**Comment by parasharrajat** [comment]
@mananjadhav Please keep us posted.

**Comment by mananjadhav** [comment]
I was OOO last week, I'll push the PR today/tomorrow max.

**Comment by mountiny** [comment]
Thanks for the update!

**Comment by mananjadhav** [comment]
I am half done with the work and updates. Focusing on the MenuItem has some issue with the border, I’ll close this over the weekend

**Comment by bfitzexpensify** [comment]
How is this one looking @mananjadhav? Anything you need a hand with?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mananjadhav** [comment]
@bfitzexpensify As this is up for payout today, fyi based on this comment https://github.com/Expensify/App/pull/7846#discussion_r961610917, we had decided to double the budget for this one due to increased effort on QA across multiple components.

cc - @roryabraham

**Comment by bfitzexpensify** [comment]
Thanks for sharing that comment @mananjadhav. I recreated the job at the higher budget and sent offers for the work and C+ review.

**Comment by mananjadhav** [comment]
Thanks for prompt action @bfitzexpensify, accepted.

**Comment by mountiny** [comment]
Thanks for pointing this out Manan and sorry we havent noted this before here.

**Comment by mananjadhav** [comment]
No worries at all @mountiny and it looks like it is paid out. Thanks @bfitzexpensify

**Comment by mountiny** [comment]
closing then, thank you everyone!

**Comment by mountiny** [comment]
Ooops @bfitzexpensify

**Comment by bfitzexpensify** [comment]
Sent it to you again @parasharrajat

**Comment by bfitzexpensify** [comment]
OK, looks like we're all good now!
